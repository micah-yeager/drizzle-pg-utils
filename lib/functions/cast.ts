import { type SQL, type SQLWrapper, sql } from "drizzle-orm"
import type {
	ColumnBuilderBase,
	ColumnBuilderBaseConfig,
	ColumnDataType,
	MakeColumnConfig,
} from "drizzle-orm/column-builder"
import {
	type AnyPgTable,
	PgColumn,
	type PgColumnBuilder,
	PgTable,
} from "drizzle-orm/pg-core"
import type { TypeFromSQL } from "../util"

/** Mimic the internal `build` method on PgColumnBuilder. */
interface PgColumnBuilder_ extends PgColumnBuilder {
	build(
		table: AnyPgTable<{ name: string }>,
	): PgColumn<
		MakeColumnConfig<ColumnBuilderBaseConfig<ColumnDataType, string>, string>
	>
}

/** A blank table to use for casting. */
const blankTable = new PgTable("", undefined, "")

/** @see https://www.postgresql.org/docs/17/sql-expressions.html#SQL-SYNTAX-TYPE-CASTS */
export function cast<T extends PgColumn | ColumnBuilderBase>(
	expression: SQLWrapper,
	type: T | ((...args: any[]) => T),
): SQL<TypeFromSQL<T, { nullable: false }>> {
	let sqlType: string
	// Column instance.
	if (type instanceof PgColumn) {
		sqlType = type.getSQLType()
	}
	// Column builder.
	else {
		const builder = (
			typeof type === "object" ? type : type()
		) as PgColumnBuilder_
		sqlType = builder.build(blankTable).getSQLType()
	}

	return sql`cast(${expression} as ${sql.raw(sqlType)})`
}
