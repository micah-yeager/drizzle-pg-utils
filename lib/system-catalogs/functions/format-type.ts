import { type SQL, type SQLWrapper, sql } from "drizzle-orm"

/** @see https://www.postgresql.org/docs/17/functions-info.html#FUNCTIONS-INFO-CATALOG-TABLE */
export function formatType(oid: SQLWrapper, typemod: SQLWrapper): SQL<string> {
	return sql`format_type(${oid}, ${typemod})`
}
