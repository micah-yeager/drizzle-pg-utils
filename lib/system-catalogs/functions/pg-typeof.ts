import { type SQL, type SQLWrapper, sql } from "drizzle-orm"

/** @see https://www.postgresql.org/docs/17/functions-info.html#FUNCTIONS-INFO-CATALOG-TABLE */
export function pgTypeof(input: SQLWrapper): SQL<string> {
	return sql`pg_typeof(${input})`
}
