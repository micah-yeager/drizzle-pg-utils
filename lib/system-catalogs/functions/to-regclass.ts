import { type SQL, type SQLWrapper, sql } from "drizzle-orm"

/** @see https://www.postgresql.org/docs/17/functions-info.html#FUNCTIONS-INFO-CATALOG-TABLE */
export function toRegclass(input: SQLWrapper | string): SQL<string | null> {
	const normalized = typeof input === "string" ? sql<string>`${input}` : input
	return sql`to_regclass(${normalized})`
}
