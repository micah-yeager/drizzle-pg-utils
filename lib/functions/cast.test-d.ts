import { type SQL, sql } from "drizzle-orm"
import { integer, json, pgTable, text } from "drizzle-orm/pg-core"
import { describe, expectTypeOf, it } from "vitest"
import { type CircleTuple, circle } from "../columns/circle"
import { money } from "../columns/money"
import { cast } from "./cast"

type TestJson = { test: string; nested: { test: number }; literal: "literal" }

it("should return as unknown if using string to cast", () => {
	expectTypeOf(cast(sql``, "text")).toEqualTypeOf<SQL<unknown>>()
})

it("should return as type if using column builder", () => {
	expectTypeOf(cast(sql``, integer())).toEqualTypeOf<SQL<number>>()
})

it("should return as type if using table column", () => {
	const table = pgTable("", { col: integer() })

	expectTypeOf(cast(sql``, table.col)).toEqualTypeOf<SQL<number>>()
})

describe("column builder function", () => {
	it("should return correctly for built-in type", () => {
		expectTypeOf(cast(sql``, integer)).toEqualTypeOf<SQL<number>>()
	})

	it("should return correctly for custom type from `customType()`", () => {
		expectTypeOf(cast(sql``, money)).toEqualTypeOf<SQL<string>>()
	})

	it("should return correctly for custom type from `PgColumnBuilder`", () => {
		expectTypeOf(cast(sql``, circle)).toEqualTypeOf<SQL<CircleTuple>>()
	})
})

describe("return type override", () => {
	it("should return correctly for unknown types", () => {
		expectTypeOf(cast<TestJson>(sql``, json)).toEqualTypeOf<SQL<TestJson>>()
	})

	it("should return correctly for inferred types", () => {
		expectTypeOf(cast<number>(sql``, text)).toEqualTypeOf<SQL<number>>()
	})
})
