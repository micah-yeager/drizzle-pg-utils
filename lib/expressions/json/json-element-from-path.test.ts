import { sql } from "drizzle-orm"
import { json } from "drizzle-orm/pg-core"
import { expect, it } from "vitest"
import { cast } from "../../functions/cast"
import {
	jsonElementFromPathArray,
	jsonElementFromPathArrayAsText,
} from "./json-element-from-path"

const testObj = { test: 0, nested: { key: "value" } } as const

it("should execute correctly with jsonElementFromPathArray", async () => {
	const toTest = jsonElementFromPathArray(
		cast(sql`${JSON.stringify(testObj)}`, json("").$type<typeof testObj>()),
		["nested", "key"],
	)

	const result = await globalThis.db.execute(sql`select ${toTest}`)

	expect(result.rows).toMatchSnapshot()
})

it("should execute correctly with jsonElementFromPathArrayAsText", async () => {
	const toTest = jsonElementFromPathArrayAsText(
		cast(sql`${JSON.stringify(testObj)}`, json("").$type<typeof testObj>()),
		["nested", "key"],
	)

	const result = await globalThis.db.execute(sql`select ${toTest}`)

	expect(result.rows).toMatchSnapshot()
})
