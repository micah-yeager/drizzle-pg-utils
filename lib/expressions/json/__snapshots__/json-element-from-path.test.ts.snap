// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should execute correctly with jsonElementFromPathArray 1`] = `
[
  {
    "?column?": "value",
  },
]
`;

exports[`should execute correctly with jsonElementFromPathArrayAsText 1`] = `
[
  {
    "?column?": "value",
  },
]
`;
